File: cpp/minigrep.cpp
/*
 * Minigrep in C++
 * Example: minigrep-cpp <pattern> <file>
 */

#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <regex>

using namespace std;

int main(int argc, char *argv[]) {
    if (argc < 3) {
        cout << "Usage: minigrep-cpp <pattern> <file>" << endl;
        return 1;
    }

    string pattern = argv[1];
    string filename = argv[2];

    ifstream file(filename);
    string line;
    int count = 0;
    bool found = false;
    regex reg(pattern);
    
    while (getline(file, line)) {
        if (regex_search(line, reg)) {
            cout << count << ": " << line << endl;
            found = true;
        }
        count++;
    }

    if (!found) {
        cout << "No matches found." << endl;
    }

    return 0;
}

File: elixir/minigrep.exs
# Minigrep in Elixir
# Run `elixir minigrep.exs <pattern> <file>`

args = System.argv()

if length(args) < 2 do
  IO.puts("Usage: minigrep-elixir <pattern> <file>")
  System.halt(1)
end

pattern = Enum.at(args, 0)
file = Enum.at(args, 1)

if !File.exists?(file) do
  IO.puts("File not found")
  System.halt(1)
end

File.stream!(file)
|> Stream.with_index()
|> Stream.filter(fn {line, _index} -> String.contains?(line, pattern) end)
|> Stream.map(fn {line, index} -> IO.puts("#{index}: #{line}") end)
|> Stream.run()
